### DISCUSSION TITLE
Is module federation coming to parcel ?

### DISCUSSION THREAD

**Discussion by dimitrisfasoulas** [parcel-bundler/parcel#5490] [Q&A]
I saw an article from @ScriptedAlchemy on dev.to back in August and in a [comment ](https://dev.to/scriptedalchemy/comment/143al)there he writes 

> As for support, I've already released partial support for rollup and parcel will be adhering to our API interface. All three bundlers should be compatible with each other since we are working to make this interface universal. No lock-in

So I wanted to ask if this is actively being worked on for parcel v2 🤞

**Comment by DeMoorJasper** [comment]
It's on the roadmap but not a priority at the moment. I don't think we were planning on implementing it exactly the same way as it has some flaws at scale, but I might be wrong. Maybe we'll implement 2 versions.

Pretty sure we want to ship a stable version before this. It's not that highly requested atm

**Reply by arshita04** [reply]
They have rolled the support for roll-up, is there a support for parcel

**Reply by DeMoorJasper** [reply]
Like I mentioned it's not a high priority at the moment and we want to ship a stable v2 first regardless

**Comment by ScriptedAlchemy** [comment]
If we can keep the basic get,init api it would be nice to offer the same capability across various bundlers . 

However if that's not offered, I imagine all one would need to do is make a little shim. I used to do this with webpack 4 and pretty much replicated the MF api for v4 &v5. We did similar with rollup as well. 

I'm not a big parcel user but I imagine that if they support some interface, it could work. 

Main challenge is each compiler needs the dependency sharing/fallback mechanism. The "exposes” and “remotes” functionality is relatively trivial to replicate regardless of bundler used

**Reply by pastelsky** [reply]
Are there any examples of the interface the shim would have to implement? 

It might be a common scenario where the product exposing modules is built by webpack 5, but the host still uses webpack 4 / parcel.  While it might not be possible to get all the benefits, it'd be good to know what will and will not work if the host had no native support for federation.

**Reply by ScriptedAlchemy** [reply]
It's possible to shim some functionality depending on what you're needing to shim

**Comment by ckken** [comment]
has progress?

**Comment by padcom** [comment]
Looks like this has been abandoned,...

**Reply by ScriptedAlchemy** [reply]
If someone reaches out to me I'm happy to collaborate. But other bundlers are very low on my priority list. If it's up to me to do, it'll be when someone funds the dev time or a core maintainer wants to pair on it.

**Comment by ScriptedAlchemy** [comment]
Found a possible way to bridge support to all other bundlers. Use a child compiler in webpack and just pass parcel options to it. 

So parcel emits all its build code into a webpack graph and delivers your app wrapped in the webpack runtime - the important part. 

Under the hood it doesn't matter who or what does code generation

**Comment by ScriptedAlchemy** [comment]
Update. Not for parcel but could be adapted. We are releasing module-federation/native which will initially support esBuild.  
It will have same feature set as webpacks, uses ESM and import maps + similar webpack runtime design for dependency negotiations. 

This should be bundler agnostic since we can now work around the weak apis of other bundlers by augmenting import map polyfill as a replacement for module factories and robust api of webpack. 

So, soon we will be able to support any bundler. 

Will take time to adapt for parcel and rollup, but is now possible as long as you use ESM and modern specs

**Reply by ScriptedAlchemy** [reply]
Once we release it. I'll be seeking support to bring this to parcel and rollup of anyone would like to help my organization.

**Reply by alex-milanov** [reply]
@ScriptedAlchemy any updates?

**Reply by ScriptedAlchemy** [reply]
No updates. The runtime has been released and I've successfully implemented it in bundlers like esbuild and vite. So if someone wants to implement the runtime... it's designed for it now

**Comment by kutyel** [comment]
How difficult would be to integrate https://github.com/module-federation/typescript into Parcel? 🤔 (cc: @prasannamestha 👀)

**Reply by ScriptedAlchemy** [reply]
Not hard. It's just using webpack hooks to fire download or build calls. 

Would be happy to take a PR or new package under my org if someone wants to drive it. I have capacity issues but could help maintain if it's for the purposes of federated types.

**Reply by prasannamestha** [reply]
Before we integrate typescript streaming into parcel; is it module-federation supported by parcel already?

I'll take a look at the source code of parcel this weekend and will update the progress here

**Reply by DesDevCharan** [reply]
Is this coming to parcel?

**Reply by kutyel** [reply]
@prasannamestha any update on this? Looks like parcel definitely does not support module federation yet, but @ScriptedAlchemy is willing to maintain such package if we put in the effort, what do you think? 😄 🙏🏻

**Comment by wihlke** [comment]
@devongovett @mischnic @ScriptedAlchemy 
What support do you need to make this happen? Module federation and parcel are great, but together they're greater!

**Reply by ScriptedAlchemy** [reply]
I'm unsure. Parcel has the best api outside of webpack. So if anyone could do it, it would be parcel. However I'm in the middle of redesigning federation. So I'd hold till that's done next Q as there will be many new capabilities in webpack and rspack federation. I'm also looking at a portable runtime where webpack runtime works like a package so you can mail you import into a createContainer runtime factory

**Comment by revskill10** [comment]
Any news ?

**Reply by ScriptedAlchemy** [reply]
You can use module-federation/runtime package - won't be able to create remotes with it tho
